[WPE] Kinetic scrolling doesn't work in overflow scrolling

psaavedra opened the pull request:
#### 431d9f7c6bc7092deb4ed41cbd61d27833e6223c
<pre>
[WPE] Kinetic scrolling doesn&apos;t work in overflow scrolling
<a href="https://bugs.webkit.org/show_bug.cgi?id=234433">https://bugs.webkit.org/show_bug.cgi?id=234433</a>

This extends the fix applied on bug 223133 &quot;[GTK] Kinetic iframe async
scrolling does not work&quot; [1] to the rest of the ports using nicosia.

For WPE (libwpe), the end of momentum scroll event can scroll contents even if
with a delta of 0. The explanation is in the implementation of the libwpe
&apos;handle_axis_event()&apos; API method in /Source/WebKit/UIProcess/API/wpe/WPEView.cpp:

   We treat an axis motion event with a value of zero to be equivalent to a &apos;stop&apos;
   event. Motion events with zero motion don&apos;t exist naturally,
   so this allows a backend to express &apos;stop&apos; events without changing API.
   The wheel event phase is adjusted accordingly.

[1] <a href="https://bugs.webkit.org/show_bug.cgi?id=223133">https://bugs.webkit.org/show_bug.cgi?id=223133</a>

Reviewed by Carlos Garcia Campos.

* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::eventCanScrollContents const):

Canonical link: <a href="https://commits.webkit.org/254099@main">https://commits.webkit.org/254099@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3177de78e78d131c9e2e863c7f23b7aedd9e30e0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88026 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/32188 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18744 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97171 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/152660 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91994 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30555 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26503 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80124 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/91917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93636 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24620 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/74676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24593 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79547 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79678 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/67576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/28188 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/28290 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/14541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2876 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/31314 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/74676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/30261 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/33788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
```diff
@@ -172,13 +172,10 @@ FloatPoint ScrollingTreeScrollingNode::maximumScrollPosition() const
-    // always have a delta of 0. More information in the documentation
-    // of the API gdk_event_is_scroll_stop_event.
+    // The end of momentum scroll event can scroll contents even if with delta of 0.
+    // More information in the documentation of the API gdk_event_is_scroll_stop_event.
```
It's weird to point to GTK documentation to explain a cross-platform change, In commit message you say this is only for ports using nicosia, but this is cross-platform, no?

psaavedra replied:
You are right. For some reason that I can't explain I was reading this code as Nicosia specific.

For the gdk_event_is_scroll_stop_event,  also you are right. Maybe @alexgcastro can do some suggestion for complementing the comment.

psaavedra replied:
Finally I found why this change is also valid for WPE. The reason why is in the  implementation of the  **libwpe** `handle_axis_event` API. In https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/API/wpe/WPEView.cpp#L189

```
        // We treat an axis motion event with a value of zero to be equivalent
        // to a 'stop' event. Motion events with zero motion don't exist naturally,
        // so this allows a backend to express 'stop' events without changing API.
        // The wheel event phase is adjusted accordingly.
        WebWheelEvent::Phase phase = WebWheelEvent::Phase::PhaseChanged;
        WebWheelEvent::Phase momentumPhase = WebWheelEvent::Phase::PhaseNone;
```



The motion events for the WPE port it comes through the UI-process using the `wpe_view_backend_dispatch_axis_event()` method of the **libwpe** API. For example, for Cog (@aperezdc):

```cpp
    struct wpe_input_axis_event event = {
        wpe_input_axis_event_type_motion,
        time,
        wl_data.pointer.x,
        wl_data.pointer.y,
        axis,
        - wl_fixed_to_int (value),
    };

    wpe_view_backend_dispatch_axis_event (wpe_view_data.backend, &event);
```

The  `wpe_view_backend_dispatch_axis_event()` method basically delegates the implementation in a backend what is implemented in `WebKit/UIProcess/API/wpe/WPEView.cpp` (WebKit):

``` cpp
void
wpe_view_backend_dispatch_axis_event(struct wpe_view_backend* backend, struct wpe_input_axis_event* event)
{
    if (backend->input_client)
        backend->input_client->handle_axis_event(backend->input_client_data, event);
}
```
in `src/view-backend.c`

... and finally, in https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/API/wpe/WPEView.cpp#L189 the motion event is interpreted artificially as an `phase::ended` event. (//CC @zdobersek )

carlosgcampos commented on Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
```diff
@@ -172,10 +172,11 @@ FloatPoint ScrollingTreeScrollingNode::maximumScrollPosition() const
+    // The end of momentum scroll event can scroll contents even if with delta of 0:
+    // - For GTK, more information in the documentation of the API gdk_event_is_scroll_stop_event.
+    // - For WPE (libwpe), an axis motion event with a value of zero is treat as a 'stop' event in
+    //   the implementation of 'handle_pointer_event()' in Source/WebKit/UIProcess/API/wpe/WPEView.cpp.
```
I would simplify this with just "In case of GTK and WPE the end of momentum scroll events always have a dela of 0." No need to point out to further info in any case.

psaavedra replied:
I don't mind For the case of GTK  but for the particular case of WPE I have doubts. It took to me time to find why this was having this behavior and this is artificially added in the code (https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/API/wpe/WPEView.cpp#L189) it doesn't come from any spec or protocol definition.


Having say that maybe it is enough just with the explanation done in the commit message.

psaavedra committed:
[WPE] Kinetic scrolling doesn't work in overflow scrolling
https://bugs.webkit.org/show_bug.cgi?id=234433

This extends the fix applied on bug 223133 "[GTK] Kinetic iframe async
scrolling does not work" [1] to the rest of the ports using nicosia.

For WPE (libwpe), the end of momentum scroll event can scroll contents even if
with a delta of 0. The explanation is in the implementation of the libwpe
'handle_axis_event()' API method in /Source/WebKit/UIProcess/API/wpe/WPEView.cpp:

   We treat an axis motion event with a value of zero to be equivalent to a 'stop'
   event. Motion events with zero motion don't exist naturally,
   so this allows a backend to express 'stop' events without changing API.
   The wheel event phase is adjusted accordingly.

[1] https://bugs.webkit.org/show_bug.cgi?id=223133

Reviewed by Carlos Garcia Campos.

* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::eventCanScrollContents const):

Canonical link: https://commits.webkit.org/254099@main
